File: divisible sum pairs/Divisible Sum Pairs.ts
import * as readline from 'readline';

function divisibleSumPairs(n: number, k: number, ar: number[]): number {
  let count = 0;

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if ((ar[i] + ar[j]) % k === 0) {
        count++;
      }
    }
  }

  return count;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLines: string[] = [];
rl.on('line', (line: string) => {
  inputLines.push(line);
});

rl.on('close', () => {
  const [n, k] = inputLines[0].split(' ').map(Number);
  const ar = inputLines[1].split(' ').map(Number);

  const result = divisibleSumPairs(n, k, ar);
  console.log(result);
});

/* 
Baris per baris, berikut adalah penjelasan kode di atas:

import * as readline from 'readline';: Ini adalah sintaks untuk mengimpor modul readline dari Node.js. Modul ini digunakan untuk membaca input dari pengguna melalui terminal.

function divisibleSumPairs(n: number, k: number, ar: number[]): number {: Ini adalah deklarasi fungsi divisibleSumPairs dengan tiga parameter: n (sebuah angka), k (sebuah angka), dan ar (sebuah array angka). Fungsi ini mengembalikan sebuah angka.

let count = 0;: Ini mendeklarasikan variabel count dan menginisialisasinya dengan nilai 0. Variabel ini akan digunakan untuk menghitung jumlah pasangan angka yang dapat dibagi habis oleh k.

for (let i = 0; i < n; i++) {: Ini adalah loop for yang akan berjalan dari i = 0 hingga i < n. Loop ini digunakan untuk memeriksa semua pasangan angka dalam array ar.

for (let j = i + 1; j < n; j++) {: Ini adalah loop for yang akan berjalan dari j = i + 1 hingga j < n. Loop ini digunakan untuk memeriksa semua pasangan angka setelah ar[i] dalam array ar.

if ((ar[i] + ar[j]) % k === 0) {: Ini adalah kondisi if yang memeriksa apakah jumlah dari ar[i] dan ar[j] dapat dibagi habis oleh k.

count++;: Jika pasangan angka tersebut dapat dibagi habis oleh k, maka count akan ditambah satu.

return count;: Fungsi divisibleSumPairs mengembalikan nilai dari variabel count, yaitu jumlah pasangan angka yang dapat dibagi habis oleh k.

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });: Membuat objek rl menggunakan fungsi createInterface dari modul readline. Objek ini akan digunakan untuk membaca input dari pengguna melalui terminal.

let inputLines: string[] = [];: Mendeklarasikan variabel inputLines sebagai array kosong. Array ini akan digunakan untuk menyimpan baris-baris input dari pengguna.

rl.on('line', (line: string) => { inputLines.push(line); });: Menambahkan event listener ke objek rl yang akan memasukkan setiap baris input pengguna ke dalam array inputLines saat pengguna menekan tombol Enter.

rl.on('close', () => {: Menambahkan event listener ke objek rl yang akan dieksekusi ketika pengguna menutup input.

const [n, k] = inputLines[0].split(' ').map(Number);: Menguraikan baris pertama dari inputLines menjadi dua angka, n dan k, menggunakan fungsi split dengan pemisah spasi. Angka-angka tersebut kemudian diubah menjadi tipe data Number menggunakan fungsi map(Number).

const ar = inputLines[1].split(' ').map(Number);: Menguraikan baris kedua dari inputLines menjadi array ar menggunakan fungsi split dengan pemisah spasi. Angka-angka tersebut kemudian diubah menjadi tipe data Number menggunakan fungsi map(Number).

const result = divisibleSumPairs(n, k, ar);: Memanggil fungsi divisibleSumPairs dengan argumen n, k, dan ar, dan menyimpan hasilnya ke dalam variabel result.

console.log(result);: Mencetak hasil yang didapatkan ke dalam terminal.
*/

File: divisible sum pairs/Divisible Sum Pairs.cs
using System;

// Menggunakan namespace System yang menyediakan kelas-kelas dan fungsi-fungsi dasar
// yang digunakan dalam pemrograman C#
class Program
{
    static int divisibleSumPairs(int n, int k, int[] ar)
    {
        int count = 0;

        // Loop untuk setiap elemen pada array
        for (int i = 0; i < n; i++)
        {
            // Loop untuk mencari pasangan elemen
            for (int j = i + 1; j < n; j++)
            {
                // Mengecek apakah penjumlahan dua elemen dapat dibagi habis oleh k
                if ((ar[i] + ar[j]) % k == 0)
                {
                    count++; // Jika ya, tambahkan count
                }
            }
        }

        return count; // Mengembalikan jumlah pasangan yang memenuhi kondisi
    }

    static void Main()
    {
        int n, k;

        // Membaca input jumlah elemen dan bilangan pembagi dari pengguna
        string[] input = Console.ReadLine().Split();
        n = int.Parse(input[0]);
        k = int.Parse(input[1]);

        int[] ar = new int[n];

        // Membaca elemen-elemen array dari pengguna
        string[] arrayInput = Console.ReadLine().Split();
        for (int i = 0; i < n; i++)
        {
            ar[i] = int.Parse(arrayInput[i]);
        }

        // Memanggil fungsi divisibleSumPairs untuk menghitung jumlah pasangan yang memenuhi kondisi
        int result = divisibleSumPairs(n, k, ar);

        // Menampilkan hasil perhitungan ke layar
        Console.WriteLine(result);
    }
}

/*
mengimplementasikan sebuah fungsi untuk mencari jumlah pasangan angka dalam sebuah array yang dapat dibagi habis oleh suatu bilangan.

Berikut adalah penjelasan per baris kode tersebut:

using System;: Ini adalah pernyataan yang memberitahu compiler untuk menggunakan namespace System. Namespace System menyediakan berbagai kelas dan fungsi yang sering digunakan dalam pemrograman C#, termasuk kelas-kelas yang terkait dengan input/output.

class Program: Membuka definisi kelas Program.

static int divisibleSumPairs(int n, int k, int[] ar): Ini adalah definisi fungsi divisibleSumPairs. Fungsi ini mengambil tiga argumen: n (jumlah elemen dalam array), k (bilangan yang harus habis dibagi), dan ar (array integer). Fungsi ini menghitung dan mengembalikan jumlah pasangan elemen dalam array ar yang dapat dibagi habis oleh k.

int count = 0;: Mendeklarasikan dan menginisialisasi variabel count dengan nilai 0. Variabel ini digunakan untuk menghitung jumlah pasangan yang memenuhi kondisi.

for (int i = 0; i < n; i++): Memulai loop for untuk variabel i dari 0 hingga n-1. Loop ini digunakan untuk mengiterasi setiap elemen dalam array.

for (int j = i + 1; j < n; j++): Memulai loop for dalam loop for sebelumnya untuk variabel j dari i + 1 hingga n-1. Loop ini digunakan untuk mencari pasangan elemen yang belum dihitung dengan elemen ar[i].

if ((ar[i] + ar[j]) % k == 0): Memeriksa apakah penjumlahan elemen ar[i] dan ar[j] dapat dibagi habis oleh k. Jika ya, maka pasangan ini memenuhi kondisi dan count akan ditambah satu.

return count;: Mengembalikan nilai count setelah selesai menghitung jumlah pasangan.

static void Main(): Definisi fungsi Main. Ini adalah titik masuk untuk eksekusi program.

int n, k;: Mendeklarasikan variabel n dan k untuk menyimpan input.

string[] input = Console.ReadLine().Split();: Membaca input dari pengguna sebagai sebuah string dan membaginya menjadi array string menggunakan spasi sebagai pemisah. Kemudian, array string tersebut disimpan dalam variabel input.

n = int.Parse(input[0]);: Mengonversi elemen pertama dalam array input menjadi integer dan menyimpannya dalam variabel n.

k = int.Parse(input[1]);: Mengonversi elemen kedua dalam array input menjadi integer dan menyimpannya dalam variabel k.

int[] ar = new int[n];: Mendeklarasikan array ar dengan ukuran n untuk menyimpan elemen-elemen input.

string[] arrayInput = Console.ReadLine().Split();: Membaca input dari pengguna sebagai sebuah string dan membaginya menjadi array string menggunakan spasi sebagai pemisah. Kemudian, array string tersebut disimpan dalam variabel arrayInput.

for (int i = 0; i < n; i++): Memulai loop for untuk variabel i dari 0 hingga n-1. Loop ini digunakan untuk mengiterasi setiap elemen dalam array.

ar[i] = int.Parse(arrayInput[i]);: Mengonversi setiap elemen dalam array arrayInput menjadi integer dan menyimpannya dalam array ar pada indeks yang sesuai.

int result = divisibleSumPairs(n, k, ar);: Memanggil fungsi divisibleSumPairs dengan argumen yang sesuai dan menyimpan hasilnya dalam variabel result.

Console.WriteLine(result);: Menampilkan hasil perhitungan ke layar.

Kode di atas membaca dua baris input dari pengguna. Baris pertama berisi dua angka yang dipisahkan oleh spasi, yaitu n (jumlah elemen dalam array) dan k (bilangan yang harus habis dibagi). Baris kedua berisi n angka yang dipisahkan oleh spasi, yaitu elemen-elemen array ar. Setelah itu, program menghitung dan menampilkan jumlah pasangan elemen dalam array ar yang dapat dibagi habis oleh k.
*/